Short-circuit `GetLegacySigOpCount` for known scripts

l0rinc opened the pull request:
It is an optimization PR for `GetSigOpCount`, covered heavily with new tests, benchmarks and small refactorings.

The changes were split out to https://github.com/bitcoin/bitcoin/pull/32729 (refactors only for now).

-------

### Context

Test coverage was thin for this code path that suddenly became popular for different reasons (https://github.com/bitcoin/bitcoin/pull/31624 and https://github.com/bitcoin/bitcoin/pull/32521 and https://github.com/bitcoin/bitcoin/pull/32533) highlighting the need for better code coverage.

The PR now splits out the common `PUSHDATA` length/bounds reads and error checks and covers its corner cases with unit tests and benchmarks.

### Testing

* added unit tests covering every standard script type (and a historical `OP_RETURN … OP_CHECKSIG` oddity);
* added fuzz-style test that compares the refactored implementation against the *exact* legacy algorithm on random scripts - also ran it continuously for 25 hours without divergence;
* reindexed to height 897,000 [comparing the outputs of the old and the new implementation](https://gist.github.com/l0rinc/ff6347164b3c502a9643ba7be171860e) without incidents.

l0rinc committed:
bench: measure `CheckBlock` speed separately from serialization

To make sure the blocks keep their original state, a ResetChecked() was introduced which will reenable all relevant calculations

l0rinc committed:
bench: measure `SigOpsBlock` speeds separately

l0rinc committed:
refactor: rename `GetSigOpCount` to `GetLegacySigOpCount`

Also added primitive argument name reminder to the call sites to highlight the meaning of the call.

l0rinc committed:
test: add `SigOpCount` edge‑cases & known‑template coverage

* `GetLegacySigOpCountErrors` feeds malformed PUSHDATA sequences to confirm the parser never crashes and still counts sig‑ops that appear before the error.
* `GetLegacySigOpCountKnownTemplates` asserts the expected legacy/accurate sig‑op totals for all common known script templates (P2PKH, P2SH, P2WPKH/WSH, P2TR, compressed & uncompressed P2PK, OP_RETURN, multisig).

l0rinc committed:
refactor: extract `DecodePushData` from `GetScriptOp` and use it in `GetLegacySigOpCount` as well

Introduce helper `DecodePushData` that parses the size prefixes, checks bounds, and returns size_bytes & data_bytes - or error on truncation.

To make absolutely sure it's just a refactor, a new unit test was added which asserts the refactored `GetLegacySigOpCount` exactly matches the previous algorithm.

tapcrafter commented on src/script/script.h:
```diff
@@ -552,8 +553,59 @@ class CScript : public CScriptBase
+        return size() == 25 &&
+               front() == OP_DUP &&
+               (*this)[1] == OP_HASH160 &&
+               (*this)[2] == 0x14 &&
```
Not sure if we can include `interpreter.h` here, but if yes, we could use `WITNESS_V0_KEYHASH_SIZE`. And the other constants too, for extra clarity (`WITNESS_V0_SCRIPTHASH_SIZE = 32`, `WITNESS_V1_TAPROOT_SIZE = 32`).

tapcrafter commented on src/script/script.h:
```diff
@@ -552,8 +553,59 @@ class CScript : public CScriptBase
+               (*this)[1] == 0x20;
+    }
+
+    bool IsCompressedPayToPubKey() const noexcept
```
Looks like this check omits the `&& (script[1] == 0x02 || script[1] == 0x03)` part that was in the original `IsToPubKey` function. Not sure if this is on purpose?

l0rinc replied:
I took these definitions from `solver.cpp` - the point here is not necessarily to fully validate the internal structure of the script templates, but rather to check if it follows a general template's layout (opcode, followed by a version, followed by a hash, etc). Validating the existing versions is done in other places, it's not important for e.g. SigOp count. If you think I'm mistaken here, please let me know.

tapcrafter commented on src/script/script.h:
```diff
@@ -552,8 +553,59 @@ class CScript : public CScriptBase
+               back() == OP_CHECKSIG;
+    }
+
+    bool IsUncompressedPayToPubKey() const noexcept
```
Same here, the original had the additional condition `&& script[1] == 0x04`.

l0rinc replied:
This wasn't checked in `MatchPayToPubkey` either.

tapcrafter reviewed (commented):
Concept ACK 432ff7e736942c6dd6fc923300b5fa5f0f94c287.

Can't fully speak to the correctness of the refactor itself, still brushing up on my C++. So can't give a full ACK but left some questions instead.

<details>

<summay>Test protocol</summary>

New benchmark test on 2e8a9e17 (first commit of PR, before refactor):

```
|           ns/sigops |            sigops/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               50.76 |       19,700,534.44 |    0.5% |      0.01 | `SigOpsBlockBench`
```


On 432ff7e736942c6dd6fc923300b5fa5f0f94c287 (last commit):

```
|           ns/sigops |            sigops/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               18.87 |       53,002,534.31 |    0.4% |      0.01 | `SigOpsBlockBench`
```

I also ran the unit tests a couple of times to excercise the randomized tests in `src/test/sigopcount_tests.cpp`.
No failures to report.

</details>

l0rinc commented on src/test/sigopcount_tests.cpp:
```diff
@@ -30,17 +30,16 @@ BOOST_FIXTURE_TEST_SUITE(sigopcount_tests, BasicTestingSetup)
     CScript s1;
-    BOOST_CHECK_EQUAL(s1.GetSigOpCount(false), 0U);
-    BOOST_CHECK_EQUAL(s1.GetSigOpCount(true), 0U);
+    BOOST_CHECK_EQUAL(s1.GetLegacySigOpCount(/*fAccurate=*/false), 0U);
```
It exercises both `GetSigOpCount` and `GetLegacySigOpCount`, so I kept it.
But the new tests can indeed be renamed, thanks!

ryanofsky reviewed (approved):
Concept ACK bde1fc1289443785e4538c757631d2fa19d5ecd8. Seems like this makes code clearer and faster without changing complexity, and adds tests.

darosior reviewed (commented):
This is a significant refactoring of tricky consensus critical routines that have essentially never been modified since Satoshi wrote them 15 years ago. The exploration is interesting but it seems ill-advised to completely overhaul this critical code for a marginal IBD speedup.

I feel bad for being stop energy here, but since i have been an advocate for us voicing our opinion instead of just ignoring PRs i'll go and own it. Concept NACK: i don't think the risk-benefit ratio of this change is worth it.

l0rinc commented:
@ryanofsky, @darosior, thank you for expressing your concerns.
I have removed the optimization part, I understand why that can be considered risky. The new rebase only contains extraction of the common script parts to avoid some work duplication and to help with understanding this part of the codebase better. It's also covered extensively with deterministic and random tests, comparing against fixed values, non-standard scripts, and completely random ones compared against the original implementation. We can reconsider the optimizations if this gets accepted.

Also note that the optimizations weren't there for IBD necessarily, but as the `checkblock` bench states:
```
// These are the two major time-sinks which happen after we have fully received
// a block off the wire, but before we can relay the block on to peers using
// compact block relay.
```

ryanofsky commented on src/test/sigopcount_tests.cpp:
```diff
@@ -28,6 +30,160 @@ Serialize(const CScript& s)
 
 BOOST_FIXTURE_TEST_SUITE(sigopcount_tests, BasicTestingSetup)
 
+BOOST_AUTO_TEST_CASE(GetLegacySigOpCountErrors)
```
In commit "test: add `SigOpCount` edge‑cases & known‑template coverage" (461779ae4903d8e86500fe5ef4d73009d80ee311)

Could consider expanding commit message a bit to say how the new tests compare to existing tests. I think: 

- GetLegacySigOpCountErrors is adding new coverage for malformed scripts that weren't checked before?
- GetLegacySigOpCountStandardScripts is adding more organized and complete coverage, but not really new coverage?

And I know you tend to push back against adding code comments, so ignore this suggestion if you don't think it is a good idea, but I'd consider moving test descriptions from the commit message:

>* `GetLegacySigOpCountErrors` feeds malformed PUSHDATA sequences to confirm the parser never crashes and still counts sig‑ops that appear before the error.
>* `GetLegacySigOpCountStandardScripts` asserts the expected legacy/accurate sig‑op totals for all common known script templates (P2PKH, P2SH, P2WPKH/WSH, P2TR, compressed & uncompressed P2PK, OP_RETURN, multisig).

into the test code itself, because IMO is useful to have 1-2 line descriptions of tests like these to help navigate and understand them.

ryanofsky commented on src/script/script.h:
```diff
@@ -552,11 +558,60 @@ class CScript : public CScriptBase
+
+    bool IsCompressedPayToPubKey() const noexcept
+    {
+        return size() == 35 &&
```
In commit "refactor: add extra-fast checks for all standard script types" (c87c7b02796d83f5c4d8a7472a2cd13df30e6a05)

Any particular reason for writing 35 instead of CPubKey::COMPRESSED_SIZE + 2 like previous code?

ryanofsky commented on src/compressor.cpp:
```diff
@@ -16,36 +16,27 @@
 {
-    if (script.size() == 35 && script[0] == 33 && script[34] == OP_CHECKSIG
-                            && (script[1] == 0x02 || script[1] == 0x03)) {
+    if (script.IsCompressedPayToPubKey()) {
```
In commit "refactor: add extra-fast checks for all standard script types" (c87c7b02796d83f5c4d8a7472a2cd13df30e6a05)

This is dropping the `script[1] == 0x02 || script[1] == 0x03))` check?

ryanofsky commented on src/compressor.cpp:
```diff
@@ -16,36 +16,27 @@
     }
-    if (script.size() == 67 && script[0] == 65 && script[66] == OP_CHECKSIG
-                            && script[1] == 0x04) {
+    if (script.IsUncompressedPayToPubKey()) {
```
In commit "refactor: add extra-fast checks for all standard script types" (c87c7b02796d83f5c4d8a7472a2cd13df30e6a05)

This is dropping the `script[1] == 0x04` check?

ryanofsky reviewed (commented):
Concept ACK 5b11b8706f14e0545f7e07c7c029e24811dc6c9d. I haven't finished reviewing but these changes seem very straightforward and the new test and benchmark coverage seems nice. Left some review comments below

l0rinc committed:
refactor: moved the script-hash-size constants to `script.h`

This will be needed in upcoming commit, to be able to use previously defined constants in `script.h`.
I made sure to check that there were no other constants with the same names.

l0rinc committed:
refactor: add `IsPayToPubKeyHash` helper to script.h

The usages in `compressor.cpp` and `solver.cpp` were also updated to use the new methods.

l0rinc committed:
refactor: add `IsCompressedPayToPubKey` helper to script.h

The usages in `compressor.cpp` and `solver.cpp` were also updated to use the new methods.
Note that compressor has additional checks as well, which are not properly exercised by the `compressed_p2pk` unit test.

l0rinc committed:
refactor: add `IsUncompressedPayToPubKey` helper to script.h

The usages in `compressor.cpp` and `solver.cpp` were also updated to use the new methods.
Note that compressor has additional checks as well, which are not properly exercised by the `compressed_p2pk` unit test.

l0rinc committed:
refactor: add `IsPayToWitnessPubKeyHash` helper to script.h

The usages in `compressor.cpp` were also updated to use the new methods.

l0rinc committed:
refactor: pull `IsPayToScriptHash` to header

Moved the implementation to the header (implicitly inline + noexcept), removed redundant `this` and parenthesis, used descriptive `front()`/`back()` methods and changed the `0x14` magic constant to descriptive `WITNESS_V0_KEYHASH_SIZE`.
See: https://learnmeabitcoin.com/technical/script/p2sh/#scriptpubkey

l0rinc committed:
refactor: pull `IsPayToWitnessScriptHash` to header

Moved the implementation to the header (implicitly inline + noexcept), removed redundant `this` and parenthesis, used descriptive `front()` method and changed the `0x20` magic constant to descriptive `WITNESS_V0_SCRIPTHASH_SIZE`.
See: https://learnmeabitcoin.com/technical/script/p2wsh/#scriptpubkey

l0rinc committed:
refactor: pull `IsPayToAnchor` to header

Moved the implementation to the header (implicitly inline + noexcept), removed redundant `this` and parenthesis, used descriptive `front()`/`back() methods.

l0rinc committed:
refactor: pull `IsPayToTaproot` to header

Moved the implementation to the header (implicitly inline + noexcept), removed redundant `this` and parenthesis, used descriptive `front()` method and changed the `0x20` magic constant to descriptive `WITNESS_V1_TAPROOT_SIZE`.

l0rinc reviewed (commented):
Thanks for the reviews, addressed more of the concerns in smaller commits, even more tests: https://github.com/bitcoin/bitcoin/compare/5b11b8706f14e0545f7e07c7c029e24811dc6c9d..0819a34565d01e241e5a3f884a508472861c96e1

@darosior, this change isn't about IBD, it was mostly about block connect time optimization originally - now it's rather meant as a cleanup and flooding it with tests and benchmarks to enable optimizations in follow-ups.

l0rinc reviewed (commented):
Thanks for the reviews, addressed more of the concerns in smaller commits, even more tests: https://github.com/bitcoin/bitcoin/compare/5b11b8706f14e0545f7e07c7c029e24811dc6c9d..0819a34565d01e241e5a3f884a508472861c96e1

@darosior, this change isn't about IBD, it was mostly about block connect time optimization originally - now it's rather meant as a cleanup and flooding it with tests and benchmarks to enable optimizations in follow-ups.

ryanofsky commented:
> I don't think we should make a major overhaul to tricky consensus-critical code that was never touched in 15 years

@darosior I almost wonder if we are looking at the same PR. The current set of changes seem pretty trivial and are certainly not a major overhaul.

@l0rinc since the general arguments darosior's making are worth taking seriously and apply to a wide variety of changes maybe including the changes made here, it could make sense to separate the concerns more by opening a different PR to only improve the benchmarks and tests and not refactor the script code.

IMO, having now looked at the script changes, I think they are trivial and not risky, and they could be merged with the normal review process and 3 acks. But a conversation about raising the standard for changes like this could be worth having here or somewhere else. I do think a more productive way of raising the standard would be to do something like increase the number of expected acks, instead of relying on people to shoot down changes individually with nacks.

In a project like this it is rightfully very hard to argue on the side of taking any risk, but we should look out for places where risks are conserved and trying to reduce them one place increases them other places. IMO, it is important for changes like this to be able to progress with sufficient review so we are not put in a position of maintaining code that we are afraid to modify and that no remaining developers will have familiarity with. They also help test our review process and build confidence that it works and catches real issues. And at least in the past, they've been useful for deploying security fixes before revealing them publicly.

A related question is how to decide what changes to spend time working on and reviewing given limited review and development bandwidth. I'm not sure there is an answer to this other than letting people individually decide for themselves, but I think if we want to shift attention different places, proactively asking for reviews and highlighting changes we want to encourage is probably also a better approach than trying to actively discourage other changes. "Don't think of a pink elephant" is not a wildly effective strategy.

darosior commented:
@ryanofsky Maybe "overhaul" isn't the right term if it implies that a change is necessarily huge, and is should have used "refactoring" instead. Regardless, my point stands.

l0rinc committed:
optimization: short-circuit `GetLegacySigOpCount` for known scripts

> build/bin/bench_bitcoin -filter='SigOpsBlockBench' --min-time=1000

|           ns/sigops |            sigops/s |    err% |      ins/sigops |      cyc/sigops |    IPC |     bra/sigops |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|               58.43 |       17,115,813.09 |    0.1% |          354.20 |          209.79 |  1.688 |          89.46 |    2.8% |     10.83 | `SigOpsBlockBench`
